source/common/compressor/compressor_impl.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include "envoy/compressor/compressor.h"
+
+#include "zlib.h"
+
+namespace Envoy {
+namespace Compressor {
+
+/**
+ * Implementation of compressor's interface.
+ */
+class CompressorImpl : public Compressor {
+public:
+  CompressorImpl();
+  ~CompressorImpl();
+
+  enum CompressionLevel {
+    default_compression = -1,
+    best = 9,
+    speed = 1,
+  };
+
+  enum CompressionStrategy {
+    filtered = 1,
+    huffman = 2,
+    rle = 3,
+    default_strategy = 4,
+  };
+
+  /**
+   * Sets buffer size for feeding data to the compressor routines.
+   * @param chunk amount of memory reserved for the compressor output default =
+   * 4096.
+   */
+  void setChunk(uint64_t chunk);
+
+  /**
+   * Used for compression initialization.
+   * @param level sets compression level. 'best' gives best compression, 'speed' gives best
+   * performance, 'default_compression' gives normal compression. default = default_compression.
+   * @param strategy sets the compression algorithm. 'default_strategy' is used for normal data,
+   * 'filtered' is used for data produced by a predictor, 'huffman' is used to enforce Huffman
+   * encoding, 'rle' is used to limit match distances to one. default = default_strategy.
+   * @param window_bits sets the size of the history buffer. Larger values result in better
+   * compression, but will use more memory. It normally ranges from 8 to 15, but larger values can
+   * be applied, e.g. 31 (gzip). Larger window_bits will result in better compression at memory
+   * cost. A negative window_bits strips off zlib header and checksum from the stream.
+   * @param memory_level sets how much memory should be allocated for the internal compression
+   * state. memory_level = 1 uses the least amount of memory but is slow and degrades compression
+   * ratio, memory_level = 9 uses max memory for optimal speed.
+   */
```

Comment:
I think calling init() is mandatory, but it isn't clearly stated here. Can you clarify in the comment?

It would probably also be better to move the comments on the enum values into the enums themselves (this ends up being a kind of hard to read big block 'o text).

Reply:
Very good point 👍